quickwit/quickwit-control-plane/src/scheduler.rs

```diff
@@ -17,37 +17,585 @@
-    pub fn new(cluster: Arc<Cluster>, metastore: Arc<dyn Metastore>) -> Self {
-        Self { cluster, metastore }
+    pub fn new(
+        cluster: Arc<Cluster>,
+        metastore: Arc<dyn Metastore>,
+        indexing_client_pool: IndexingClientPool,
+    ) -> Self {
+        Self {
+            cluster,
+            metastore,
+            indexing_client_pool,
+            state: IndexingSchedulerState::default(),
+        }
+    }
+
+    async fn schedule_indexing_plan(&mut self) -> anyhow::Result<()> {
+        let indexers = self.get_indexers_from_cluster_state().await;
+        if indexers.is_empty() {
+            warn!("No indexer available, cannot build a physical indexing plan.");
+            return Ok(());
+        };
+        let indexes_metadatas = self.metastore.list_indexes_metadatas().await?;
+        let source_configs: HashMap<IndexSourceId, SourceConfig> = indexes_metadatas
+            .into_iter()
+            .flat_map(|index_metadata| {
+                index_metadata
+                    .sources
+                    .into_iter()
+                    .map(move |(source_id, source_config)| {
+                        (
+                            IndexSourceId {
+                                index_id: index_metadata.index_config.index_id.to_string(),
+                                source_id,
+                            },
+                            source_config,
+                        )
+                    })
+            })
+            .collect();
+        let indexing_tasks = build_indexing_plan(&indexers, &source_configs);
+        let new_physical_plan =
+            build_physical_indexing_plan(&indexers, &source_configs, indexing_tasks);
+        self.apply_physical_indexing_plan(&indexers, new_physical_plan)
+            .await;
+        Ok(())
+    }
+
+    /// Checks if the last applied plan corresponds to the running indexing tasks present in the
+    /// chitchat cluster state. If true, do nothing.
+    /// If false, schedule a new indexing plan.
+    async fn control_running_plan(&mut self) -> anyhow::Result<()> {
+        let last_applied_plan =
+            if let Some(last_applied_plan) = self.state.last_applied_physical_plan.as_ref() {
+                last_applied_plan
+            } else {
+                return Ok(());
+            };
+
+        // TODO(fmassot): once the actor framekwork can correctly simulate time advance, change
+        // `MIN_DURATION_BETWEEN_SCHEDULING_SECS` value for test and correclty test it. Currently, I
+        // put its value 0 to make test simple.
+        if let Some(last_applied_plan_timestamp) = self.state.last_applied_plan_timestamp.as_ref() {
+            if (OffsetDateTime::now_utc().unix_timestamp() - last_applied_plan_timestamp)
+                < MIN_DURATION_BETWEEN_SCHEDULING_SECS as i64
+            {
+                return Ok(());
+            }
+        }
+        let indexers = self.get_indexers_from_cluster_state().await;
```

Comment:
can you add the type of indexers here? 
```suggestion
        let indexers = self.get_indexers_from_cluster_state().await;
```